dcicutils/validation_utils.py

```diff
@@ -0,0 +1,290 @@
+    #         schema_names = [data_type for data_type in data]
+    #         schemas = fetch_relevant_schemas(schema_names, portal_vapp=portal_vapp)
+    #     except Exception as e:
+    #         errors.append({"exception": f"Exception fetching relevant schemas: {get_error_message(e)}"})
+    #         schemas = {}
+
+    errors = []
+    schemas = schema_manager.fetch_relevant_schemas(list(data.keys()))
+
+    for data_type in data:
+        schema = schemas.get(data_type)
+        if not schema:
+            errors.append({"error": f"No schema found for: {data_type}"})
+            continue
+        data_errors = validate_data_items_against_schemas(data[data_type], data_type, schema)
+        errors.extend(data_errors)
+    return {"errors": errors} if errors else None
+
+
+def validate_data_items_against_schemas(data_items: SheetData, data_type: str, schema: JsonSchema) -> List[Dict]:
+    """
+    Like validate_data_against_schemas but for a simple list of data items each of the same given data type.
+    """
+    errors = []
+    for data_item_index, data_item in enumerate(data_items):
+        data_item_errors = validate_data_item_against_schemas(data_item, data_type, data_item_index, schema)
+        errors.extend(data_item_errors)
+    return errors
+
+
+def validate_data_item_against_schemas(data_item: AnyJsonData, data_type: str,
+                                       data_item_index: Optional[int], schema: JsonSchema) -> List[Dict]:
+    """
+    Like validate_data_against_schemas but for a single data item of the given data type.
+    The given data item index is just for informational purposes; it corresponds to the
+    ordinal index of the data item in its containing list. Uses the standard jsonschema
+    package to do the heavy lifting of actual schema validation, but exerts extra effort to
+    specifically itemize/aggregate the most common (missing and extraneous properties) errors.
+    """
+    errors = []
+
+    identifying_properties = schema.get("identifyingProperties", [])
+    identifying_value = SchemaManager.identifying_value(data_item, identifying_properties)
+    if not identifying_value:
+        errors.append({
+            "type": data_type,
+            "unidentified": True,
+            "index": data_item_index,
+            "identifying_properties": identifying_properties
+        })
+
+    def extract_single_quoted_strings(message: str) -> List[str]:
+        return re.findall(r"'(.*?)'", message)
+
+    schema_validator = jsonschema.Draft7Validator(schema)
+    for schema_validation_error in schema_validator.iter_errors(data_item):
+        if schema_validation_error.validator == "required":
+            errors.append({
+                "type": data_type,
+                "item" if identifying_value else "unidentified": identifying_value if identifying_value else True,
+                "index": data_item_index,
+                "missing_properties": schema_validation_error.validator_value})
```

Comment:
I discovered this is actually not quite right; the list of missing properties can be gotten like this (I can make this change later if you want) ...` "missing_properties": list(set(schema_validation_error.validator_value) - set(schema_validation_error.instance))`